# 05_231211_RePRSTrials

### 主目的

Tonga集団のサンプルを用いて倹約遺伝子仮説を検討する。

1. 過去のPRSを推定する：今回はその前段階として現在ベースでPRSを推定する。
2. ポリジェニックセレクションの検討：これから（まず手法をもう一度チェック！）

### 実行にあたって

Tonga集団のデータを用いてPRSを計算したい。が、いま手元に彼らの身長や体重に関するデータがないことがわかった。

とりあえずその前段階として遺伝統計学夏の学校2023の実習を再現してみる。

### 遺伝統計学夏の学校2023【QC以降の手順確認】

まずはデータファイルをダウンロード・解凍する。

In [12]:
mkdir -p ~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
cd ~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
ls

In [13]:
wget http://www.sg.med.osaka-u.ac.jp/files/StatGenSummerSchool2023_GenomeDataAnalysis5.zip
ls

--2023-12-18 16:22:57--  http://www.sg.med.osaka-u.ac.jp/files/StatGenSummerSchool2023_GenomeDataAnalysis5.zip
Resolving www.sg.med.osaka-u.ac.jp (www.sg.med.osaka-u.ac.jp)... 133.1.138.171
Connecting to www.sg.med.osaka-u.ac.jp (www.sg.med.osaka-u.ac.jp)|133.1.138.171|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 185582439 (177M) [application/zip]
Saving to: ‘StatGenSummerSchool2023_GenomeDataAnalysis5.zip’

StatGenSummerSchool 100%[===================>] 176.98M  21.7MB/s    in 6.7s    

2023-12-18 16:23:04 (26.3 MB/s) - ‘StatGenSummerSchool2023_GenomeDataAnalysis5.zip’ saved [185582439/185582439]

StatGenSummerSchool2023_GenomeDataAnalysis5.zip


In [14]:
unzip StatGenSummerSchool2023_GenomeDataAnalysis5.zip

Archive:  StatGenSummerSchool2023_GenomeDataAnalysis5.zip
   creating: StatGenSummerSchool2023_GenomeDataAnalysis5/
   creating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/
   creating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/


  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/1KG_EUR_QC.bed  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/1KG_EUR_QC.bim  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/1KG_EUR_QC.fam  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/phenotype_RA.txt  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/RA_GWASmeta_European_PRSise.txt.gz  
   creating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS-PRSboxplot.png  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS.best  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS.best.txt  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS.log  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS.

In [15]:
ls

StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip


In [16]:
ls StatGenSummerSchool2023_GenomeDataAnalysis5

PRSice  StatGenSummerSchool2023_GenomeDataAnalysis5.pdf


In [17]:
ls StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice

Input                      PRsice_Command.txt  PRSice_mac.R
Output                     PRSice_linux        PRSice.R
PRsice_Command_cygwin.txt  PRSice_linux.R      PRSice_win64.exe
PRsice_Command_mac.txt     PRSice_mac


In [18]:
ls StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input
ls StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output

1KG_EUR_QC.bed  1KG_EUR_QC.fam    RA_GWASmeta_European_PRSise.txt.gz
1KG_EUR_QC.bim  phenotype_RA.txt


1KG_EUR-RAGWAS_BARPLOT_2019-08-25.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS_BARPLOT_2021-07-14.png        1KG_EUR-RAGWAS-PRSboxplot.png
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2019-08-25.png  1KG_EUR_Sample.txt
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2021-07-14.png  PlotPRSice.R


#### 入力ファイルの確認

In [19]:
ls StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input

1KG_EUR_QC.bed  1KG_EUR_QC.fam    RA_GWASmeta_European_PRSise.txt.gz
1KG_EUR_QC.bim  phenotype_RA.txt


In [20]:
zcat StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/RA_GWASmeta_European_PRSise.txt.gz | head -n 10

SNP	CHR	BP	A1	A2	P	OR
chr1:751343	1	751343	A	T	0.01	0.85
chr1:751756	1	751756	T	C	0.01	1.17
rs3094315	1	752566	A	G	0.0093	1.14
rs3131972	1	752721	A	G	0.009	0.88
rs3131971	1	752894	T	C	0.01	0.87
chr1:753405	1	753405	A	C	0.01	1.17
chr1:753425	1	753425	T	C	0.0097	0.87
rs2073814	1	753474	G	C	0.009	1.14
rs2073813	1	753541	A	G	0.0095	0.85


In [21]:
cat StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/phenotype_RA.txt | head -n 10

HG00096	HG00096	2
HG00097	HG00097	1
HG00099	HG00099	1
HG00100	HG00100	1
HG00101	HG00101	2
HG00102	HG00102	2
HG00103	HG00103	1
HG00104	HG00104	1
HG00106	HG00106	1
HG00108	HG00108	1


#### PRS解析の実行

In [22]:
cd ~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials/StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice
mkdir -p Output_2
touch Output_2/.gitkeep
ls

Input     PRsice_Command_cygwin.txt  PRSice_linux    PRSice_mac.R
Output    PRsice_Command_mac.txt     PRSice_linux.R  PRSice.R
Output_2  PRsice_Command.txt         PRSice_mac      PRSice_win64.exe


In [23]:
Rscript PRSice_linux.R \
--prsice PRSice_linux \
--dir . \
--out ./Output_2/1KG_EUR-RAGWAS \
--base ./Input/RA_GWASmeta_European_PRSise.txt.gz \
--target ./Input/1KG_EUR_QC \
--pheno ./Input/phenotype_RA.txt \
--thread 1 \
--stat OR \
--binary-target T

Trying to install data.table in ./lib
Trying to install optparse in ./lib
PRSice 2.3.3 (2020-08-05) 
https://github.com/choishingwan/PRSice
(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly
GNU General Public License v3
If you use PRSice in any published work, please cite:
Choi SW, O'Reilly PF.
PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.
GigaScience 8, no. 7 (July 1, 2019)
2023-12-18 16:23:34
./PRSice_linux \
    --a1 A1 \
    --a2 A2 \
    --bar-levels 0.001,0.05,0.1,0.2,0.3,0.4,0.5,1 \
    --base ./Input/RA_GWASmeta_European_PRSise.txt.gz \
    --binary-target T \
    --bp BP \
    --chr CHR \
    --clump-kb 250kb \
    --clump-p 1.000000 \
    --clump-r2 0.100000 \
    --interval 5e-05 \
    --lower 5e-08 \
    --num-auto 22 \
    --or  \
    --out ./Output_2/1KG_EUR-RAGWAS \
    --pheno ./Input/phenotype_RA.txt \
    --pvalue P \
    --seed 2986104432 \
    --snp SNP \
    --stat OR \
    --target ./Input/1KG_EUR_QC \
    --thread 1 \
    --upper 0.5

Init

#### Output結果の確認

In [24]:
ls Output_2

1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png  1KG_EUR-RAGWAS.summary


In [25]:
head Output_2/1KG_EUR-RAGWAS.prsice

Pheno	Set	Threshold	R2	P	Coefficient	Standard.Error	Num_SNP
-	Base	5e-08	0.0122466	0.0629842	12.6059	6.77991	257
-	Base	5.005e-05	0.0126982	0.0583659	22.8194	12.055	479
-	Base	0.00010005	0.01382	0.0484141	27.829	14.0998	561
-	Base	0.00015005	0.0139415	0.0474483	30.2877	15.2792	611
-	Base	0.00020005	0.0143518	0.0443525	33.0976	16.4604	658
-	Base	0.00025005	0.0146751	0.0420519	35.1525	17.2909	691
-	Base	0.00030005	0.015445	0.0370909	37.9226	18.1904	729
-	Base	0.00035005	0.0160981	0.0333494	41.2045	19.3644	778
-	Base	0.00040005	0.0162487	0.0325501	43.4408	20.3223	817


In [26]:
head Output_2/1KG_EUR-RAGWAS.summary

Phenotype	Set	Threshold	PRS.R2	Full.R2	Null.R2	Prevalence	Coefficient	Standard.Error	P	Num_SNP
-	Base	0.46	0.0512432	0.0512432	0	-	5081.42	1356.18	0.000179058	101385


In [27]:
#画像表示準備
mkdir -p ~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials/pic
cp Output_2/*.png ~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials/pic/

【1KG_EUR-RAGWAS_BARPLOT_2023-12-11.png（左）】【1KG_EUR-RAGWAS_BARPLOT_2023-12-11.png（右）】

<img src="pic/1KG_EUR-RAGWAS_BARPLOT_2023-12-11.png" width="49%">
<img src="pic/1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-11.png" width="49%">

#### PRS解析結果の検討

In [28]:
head Output_2/1KG_EUR-RAGWAS.best

FID IID In_Regression PRS
HG00096 HG00096 Yes -0.000627027476
HG00097 HG00097 Yes -0.00058856461
HG00099 HG00099 Yes -0.000713382756
HG00100 HG00100 Yes -0.000571751231
HG00101 HG00101 Yes -0.000530233701
HG00102 HG00102 Yes -0.000583765514
HG00103 HG00103 Yes -0.000573780098
HG00104 HG00104 Yes -0.000506927325
HG00106 HG00106 Yes -0.000537212649


In [29]:
#ファイルの調整
awk '$1!="FID" && $3="Yes" {print $1"\t"$2"\t"$4}' Output_2/1KG_EUR-RAGWAS.best > Output_2/1KG_EUR-RAGWAS.best.txt

In [30]:
cat Output_2/1KG_EUR-RAGWAS.best.txt | head -n 10

HG00096	HG00096	-0.000627027476
HG00097	HG00097	-0.00058856461
HG00099	HG00099	-0.000713382756
HG00100	HG00100	-0.000571751231
HG00101	HG00101	-0.000530233701
HG00102	HG00102	-0.000583765514
HG00103	HG00103	-0.000573780098
HG00104	HG00104	-0.000506927325
HG00106	HG00106	-0.000537212649
HG00108	HG00108	-0.000588234975


In [31]:
ls *

PRsice_Command_cygwin.txt  PRSice_linux    PRSice_mac.R
PRsice_Command_mac.txt     PRSice_linux.R  PRSice.R
PRsice_Command.txt         PRSice_mac      PRSice_win64.exe

Input:
1KG_EUR_QC.bed  1KG_EUR_QC.fam    RA_GWASmeta_European_PRSise.txt.gz
1KG_EUR_QC.bim  phenotype_RA.txt

lib:
data.table  getopt  optparse

Output:
1KG_EUR-RAGWAS_BARPLOT_2019-08-25.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS_BARPLOT_2021-07-14.png        1KG_EUR-RAGWAS-PRSboxplot.png
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2019-08-25.png  1KG_EUR_Sample.txt
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2021-07-14.png  PlotPRSice.R

Output_2:
1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png


In [32]:
#Rスクリプトの調整
cat Output/PlotPRSice.R -n

     1	#ディレクトリの移動
     2		rm(list = ls(all = TRUE))
     3		setwd("C:/SummerSchool/GenomeDataAnalysis5/PRSice/Output");
     4	
     5	
     6		PRS <- read.table("1KG_EUR-RAGWAS.best.txt", header=FALSE)[,3];
     7		Pop <- read.table("1KG_EUR_Sample.txt", header=FALSE)[,2];
     8	
     9		png(file="1KG_EUR-RAGWAS-PRSboxplot.png", width = 600, height = 600);
    10		par(mar = c(5, 5, 0.5, 0.5));
    11		boxplot(PRS~Pop, col=c(2,3,4,5,6), pch=16, xlab="1KG EUR sub-populations", ylab="Polygenic risk score for rheumatoid arthritis", cex.lab=2.0, cex.axis=2.0, cex.main=2.0);
    12		dev.off();
    13		
    14		anova(aov(PRS~Pop)) # P-value for Pop < 2.2x10^-16
    15		


In [33]:
cat Output/PlotPRSice.R | sed 1,5d

	PRS <- read.table("1KG_EUR-RAGWAS.best.txt", header=FALSE)[,3];
	Pop <- read.table("1KG_EUR_Sample.txt", header=FALSE)[,2];

	png(file="1KG_EUR-RAGWAS-PRSboxplot.png", width = 600, height = 600);
	par(mar = c(5, 5, 0.5, 0.5));
	boxplot(PRS~Pop, col=c(2,3,4,5,6), pch=16, xlab="1KG EUR sub-populations", ylab="Polygenic risk score for rheumatoid arthritis", cex.lab=2.0, cex.axis=2.0, cex.main=2.0);
	dev.off();
	
	anova(aov(PRS~Pop)) # P-value for Pop < 2.2x10^-16
	


In [34]:
echo 'setwd("Output_2")' > Output_2/PlotPRSice.R
cat Output/PlotPRSice.R | sed 1,5d >> Output_2/PlotPRSice.R
ls Output_2

1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png  PlotPRSice.R


In [35]:
cat Output_2/PlotPRSice.R

setwd("Output_2")
	PRS <- read.table("1KG_EUR-RAGWAS.best.txt", header=FALSE)[,3];
	Pop <- read.table("1KG_EUR_Sample.txt", header=FALSE)[,2];

	png(file="1KG_EUR-RAGWAS-PRSboxplot.png", width = 600, height = 600);
	par(mar = c(5, 5, 0.5, 0.5));
	boxplot(PRS~Pop, col=c(2,3,4,5,6), pch=16, xlab="1KG EUR sub-populations", ylab="Polygenic risk score for rheumatoid arthritis", cex.lab=2.0, cex.axis=2.0, cex.main=2.0);
	dev.off();
	
	anova(aov(PRS~Pop)) # P-value for Pop < 2.2x10^-16
	


In [36]:
#必要ファイルの準備
cp Output/1KG_EUR_Sample.txt Output_2/

In [37]:
#Rスクリプトの実行
Rscript Output_2/PlotPRSice.R

null device 
          1 
Analysis of Variance Table

Response: PRS
           Df     Sum Sq    Mean Sq F value    Pr(>F)    
Pop         4 6.4407e-07 1.6102e-07  33.237 < 2.2e-16 ***
Residuals 376 1.8215e-06 4.8440e-09                      
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


In [38]:
ls Output_2

1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        1KG_EUR-RAGWAS-PRSboxplot.png
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png  1KG_EUR_Sample.txt
1KG_EUR-RAGWAS.log                           PlotPRSice.R


In [39]:
#画像表示の準備
cp Output_2/1KG_EUR-RAGWAS-PRSboxplot.png ~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials/pic

<img src="pic/1KG_EUR-RAGWAS-PRSboxplot.png" width="49%">

### 東アジア集団GWAS（BaseData）のQC

まずややこしいのでコード格納ディレクトリとストレージディレクトリをそれぞれ変数にあてておく

In [40]:
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
ls $gitdir

strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
ls $strdir

05_231211_RePRSTrials.ipynb  Documents  jupyter_venv5  pic  test.ipynb
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip


In [41]:
cd $strdir
mkdir -p Basedata
ls

Basedata
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip


#### データのダウンロード

論文はこれを使う。

https://www-nature-com.utokyo.idm.oclc.org/articles/s41588-017-0011-x

In [42]:
cd $strdir/Basedata
wget https://portals.broadinstitute.org/collaboration/giant/images/7/78/BMI_Eastern_Asian.fmt.gzip

--2023-12-18 16:24:36--  https://portals.broadinstitute.org/collaboration/giant/images/7/78/BMI_Eastern_Asian.fmt.gzip
Resolving portals.broadinstitute.org (portals.broadinstitute.org)... 69.173.68.142
Connecting to portals.broadinstitute.org (portals.broadinstitute.org)|69.173.68.142|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12268615 (12M)
Saving to: ‘BMI_Eastern_Asian.fmt.gzip’

BMI_Eastern_Asian.f 100%[===================>]  11.70M  5.43MB/s    in 2.2s    

2023-12-18 16:24:39 (5.43 MB/s) - ‘BMI_Eastern_Asian.fmt.gzip’ saved [12268615/12268615]



In [43]:
ls $strdir/Basedata

BMI_Eastern_Asian.fmt.gzip


In [44]:
head BMI_Eastern_Asian.fmt.gzip

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	A:0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	G:0	1.6	1	0.1
10	100010909	C	T	rs141069497	T:0.0139	T:0.006614	-0.022	0.1	0.83
10	100011387	C	T	rs199733612	-	T:0,G:0	1.1	1	0.27
10	100011442	C	T	rs74153530	T:0	T:0,G:0	0.66	0.59	0.26
10	100013325	C	T	rs145270262	-	T:0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	T:0	NA	Inf	NA
10	100013382	A	G	rs141995277	G:0	G:0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	G:0	0.41	1	0.68


#### QCのチェックリスト
* Heritability Check
* Effect Allele
* File Transfer
* Genome Build
* Standard GWAS QC
* Mismatching SNPs
* Duplicate SNPs
* Ambiguous SNPs
* Sex Chromosomes
* Sample overlap
* Relatedness

##### Heritability Check

今回はスキップさせてくれ…　たぶん大丈夫だとは思うのだが。

（この前提が崩れるとそもそもこれが使えない。そういうものは公開されてないと思う。）

##### Effect Allele

MAFでフィルタリングするときに同時にチェックします。

##### File Transfer

ファイルの損傷チェック。とりあえずスキップ。

##### Genome Build：hg19/hg38が合っているかのチェック

In [45]:
head -n 2 BMI_Eastern_Asian.fmt.gzip

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	A:0	NA	Inf	NA


https://www.ncbi.nlm.nih.gov/snp/?term=rs148923672

上記を確認すると、

    10:98244118 (GRCh38)
    10:100003875 (GRCh37)
    
とあるので、このデータはGRCh37で間違いなさそう。

##### Standard GWAS QC：MAFやINFOが低いSNPは除去した方が良い！

In [46]:
cd $strdir/Basedata

In [47]:
head BMI_Eastern_Asian.fmt.gzip | \
sed -e s/A://g -e s/T://g -e s/G://g -e s/C://g | \
grep -v ,

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	0	1.6	1	0.1
10	100010909	C	T	rs141069497	0.0139	0.006614	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	0	NA	Inf	NA
10	100013382	A	G	rs141995277	0	0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	0	0.41	1	0.68


In [48]:
head BMI_Eastern_Asian.fmt.gzip | \
sed -e s/A://g -e s/T://g -e s/G://g -e s/C://g | \
grep -v , | \
awk 'NR==1 || ($6 > 0.01) {print}'

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100010909	C	T	rs141069497	0.0139	0.006614	-0.022	0.1	0.83


In [49]:
head BMI_Eastern_Asian.fmt.gzip | \
grep -v , 

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	A:0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	G:0	1.6	1	0.1
10	100010909	C	T	rs141069497	T:0.0139	T:0.006614	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	T:0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	T:0	NA	Inf	NA
10	100013382	A	G	rs141995277	G:0	G:0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	G:0	0.41	1	0.68


In [50]:
head BMI_Eastern_Asian.fmt.gzip | \
grep -v , | \
sed -e s/:/"\t"/g -e s/-/-"\t"-/g

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	-	A	0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	-	G	0	1.6	1	0.1
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	-	T	0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	-	T	0	NA	Inf	NA
10	100013382	A	G	rs141995277	G	0	G	0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	-	G	0	0.41	1	0.68


In [51]:
echo -e 'CHR\tPOS\tREF\tALT\tSNPNAME\tEAS\tEAS_MAF\tExAC_EAS\tExAC_EAS_MAF\tbeta\tse\tPvalue'
head BMI_Eastern_Asian.fmt.gzip | \
grep -v , | \
sed -e s/:/"\t"/g -e s/-/-"\t"-/g

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue


CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	-	A	0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	-	G	0	1.6	1	0.1
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	-	T	0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	-	T	0	NA	Inf	NA
10	100013382	A	G	rs141995277	G	0	G	0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	-	G	0	0.41	1	0.68


In [52]:
echo -e 'CHR\tPOS\tREF\tALT\tSNPNAME\tEAS\tEAS_MAF\tExAC_EAS\tExAC_EAS_MAF\tbeta\tse\tPvalue'
head BMI_Eastern_Asian.fmt.gzip | \
grep -v , | \
sed -e s/:/"\t"/g -e s/-/-"\t"-/g | \
sed 1d

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue


10	100003875	G	A	rs148923672	-	-	A	0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	-	G	0	1.6	1	0.1
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	-	T	0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	-	T	0	NA	Inf	NA
10	100013382	A	G	rs141995277	G	0	G	0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	-	G	0	0.41	1	0.68


In [53]:
echo -e 'CHR\tPOS\tREF\tALT\tSNPNAME\tEAS\tEAS_MAF\tExAC_EAS\tExAC_EAS_MAF\tbeta\tse\tPvalue' > BMI_Eastern_Asian.newcol.txt
cat BMI_Eastern_Asian.fmt.gzip | \
grep -v , | \
sed -e s/:/"\t"/g -e s/-/-"\t"-/g | \
sed 1d >> BMI_Eastern_Asian.newcol.txt 

In [54]:
cat BMI_Eastern_Asian.newcol.txt | \
awk 'NR==1 || ($7 > 0.01) {print}' | wc

  31194  389695 1870772


In [55]:
cat BMI_Eastern_Asian.newcol.txt | \
awk 'NR==1 || ($7 > 0.01) && ($4 == $6) {print}' | wc

  31170  389376 1869585


In [56]:
cat BMI_Eastern_Asian.newcol.txt | \
awk 'NR==1 || ($7 > 0.01) && ($4 == $6) {print}' > BMI_Eastern_Asian.txt

In [57]:
head BMI_Eastern_Asian.txt

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100017453	T	G	rs1983864	G	0.6081	G	0.5682	-	-0.031	0.015	0.044
10	100017532	G	A	rs77858766	A	0.0169	A	0.01781	-	-0.016	0.058	0.79
10	100017553	G	T	rs11189525	T	0.0129	T	0.007404	-	-0.014	0.13	0.91
10	100020880	C	T	rs33995374	T	0.1062	T	0.1163	0.077	0.061	0.21
10	100044136	G	A	rs4919199	A	0.9544	-	-	0.0076	0.036	0.83
10	100144782	C	T	rs2296441	T	0.4425	T	0.42	0.012	0.015	0.42
10	100148004	T	C	rs2182169	C	0.5615	-	-	0.014	0.015	0.37
10	100148176	A	G	rs2147896	G	0.5833	G	0.5596	0.0052	0.015	0.73


In [58]:
ls -l BMI_Eastern_Asian.txt
wc BMI_Eastern_Asian.txt

-rw-rw-r-- 1 hmiwa hmiwa 1869585 12月 18 16:24 BMI_Eastern_Asian.txt


  31170  389376 1869585 BMI_Eastern_Asian.txt


##### Mismatching SNPs：ターゲットデータで対立遺伝子が一致しているかどうかのチェック。

ターゲットデータがないとチェックできないのでここはスキップ。

##### Duplicate SNPs

In [59]:
cat BMI_Eastern_Asian.txt |\
awk '{seen[$5]++; if(seen[$5]==1){print}}' > BMI_Eastern_Asian.nodup.txt

In [60]:
ls -l BMI_Eastern_Asian.nodup.txt
wc BMI_Eastern_Asian.nodup.txt

-rw-rw-r-- 1 hmiwa hmiwa 1869585 12月 18 16:24 BMI_Eastern_Asian.nodup.txt


  31170  389376 1869585 BMI_Eastern_Asian.nodup.txt


##### Ambiguous SNPs

In [61]:
head BMI_Eastern_Asian.nodup.txt

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100017453	T	G	rs1983864	G	0.6081	G	0.5682	-	-0.031	0.015	0.044
10	100017532	G	A	rs77858766	A	0.0169	A	0.01781	-	-0.016	0.058	0.79
10	100017553	G	T	rs11189525	T	0.0129	T	0.007404	-	-0.014	0.13	0.91
10	100020880	C	T	rs33995374	T	0.1062	T	0.1163	0.077	0.061	0.21
10	100044136	G	A	rs4919199	A	0.9544	-	-	0.0076	0.036	0.83
10	100144782	C	T	rs2296441	T	0.4425	T	0.42	0.012	0.015	0.42
10	100148004	T	C	rs2182169	C	0.5615	-	-	0.014	0.015	0.37
10	100148176	A	G	rs2147896	G	0.5833	G	0.5596	0.0052	0.015	0.73


In [62]:
cat BMI_Eastern_Asian.nodup.txt |\
awk '!( ($3=="A" && $4=="T") || \
        ($3=="T" && $4=="A") || \
        ($3=="G" && $4=="C") || \
        ($3=="C" && $4=="G")) {print}' > BMI_Eastern_Asian.QC.txt

In [63]:
ls -l BMI_Eastern_Asian.QC.txt
wc BMI_Eastern_Asian.QC.txt

-rw-rw-r-- 1 hmiwa hmiwa 1707833 12月 18 16:24 BMI_Eastern_Asian.QC.txt


  28519  356256 1707833 BMI_Eastern_Asian.QC.txt


##### Sex Chromosomes

やりようがないのでスルー

##### Sample overlap
オセアニア集団とは絶対重ならないのでOK

##### Relatedness

本来はやるべきだが今回は一旦スキップ

### Tonga集団サンプル（TargetData）のQC

そもそもデータはここにあります

In [2]:
#必要な変数チェック
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
echo $gitdir
ls $gitdir

strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
echo $strdir
ls $strdir

#ストレージに移動
cd $strdir
echo $strdir
ls

/home/hmiwa/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
05_231211_RePRSTrials.ipynb  Documents  jupyter_venv5  pic  test.ipynb
/home/hmiwa/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
Basedata  StatGenSummerSchool2023_GenomeDataAnalysis5      targetdata
out.log   StatGenSummerSchool2023_GenomeDataAnalysis5.zip
/home/hmiwa/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
Basedata  StatGenSummerSchool2023_GenomeDataAnalysis5      targetdata
out.log   StatGenSummerSchool2023_GenomeDataAnalysis5.zip


In [68]:
rawdir=~/hmiwa1/rawdata/Oceania
ls $rawdir

G05_sorted.genome.vcf.gz       T129_sorted.genome.vcf.gz.tbi
G05_sorted.genome.vcf.gz.tbi   T134_sorted.genome.vcf.gz
G06_sorted.genome.vcf.gz       T134_sorted.genome.vcf.gz.tbi
G06_sorted.genome.vcf.gz.tbi   t138_sorted.genome.vcf.gz
G135_sorted.genome.vcf.gz      t138_sorted.genome.vcf.gz.tbi
G135_sorted.genome.vcf.gz.tbi  T150_sorted.genome.vcf.gz
G150_sorted.genome.vcf.gz      T150_sorted.genome.vcf.gz.tbi
G150_sorted.genome.vcf.gz.tbi  T153_sorted.genome.vcf.gz
G152_sorted.genome.vcf.gz      T153_sorted.genome.vcf.gz.tbi
G152_sorted.genome.vcf.gz.tbi  T156_sorted.genome.vcf.gz
G164_sorted.genome.vcf.gz      T156_sorted.genome.vcf.gz.tbi
G164_sorted.genome.vcf.gz.tbi  t159_sorted.genome.vcf.gz
G217_sorted.genome.vcf.gz      t159_sorted.genome.vcf.gz.tbi
G217_sorted.genome.vcf.gz.tbi  T15_sorted.genome.vcf.gz
G220_sorted.genome.vcf.gz      T15_sorted.genome.vcf.gz.tbi
G220_sorted.genome.vcf.gz.tbi  t160_sorted.genome.vcf.gz
G256_sorted.genome.vcf.gz      t160_sorted.genome.vcf.gz.t

#### QCのチェックリスト

* Sample size
* File transfer
* Genome build
* Standard GWAS QC
* Ambiguous SNPs
* Mismatching SNPs
* Duplicate SNPs
* Sex Chromosomes
* Sample Overlap
* Relatedness
* Generate final QC'ed target data file

In [65]:
mkdir -p targetdata

##### Sample size

PRS分析は本来100人以上のターゲットデータが望ましいが、残念ながら20人くらいしかいない。致し方あるまい。

##### File transfer

md5sumコードに関しては今回は設定していない。次回以降準備しておいた方が良いかもね。

##### Genome build

これはhg19でタイピングされている。Basedataと一致している。

##### Standard GWAS QC

まずはトンガ人サンプルのQCを行う。

In [85]:
ls
ls $strdir

Basedata
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip
targetdata


Basedata
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip
targetdata


In [71]:
ls $rawdir/T*vcf.gz
ls $rawdir/t*vcf.gz

/home/hmiwa/hmiwa1/rawdata/Oceania/T03_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T104_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T105_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T106_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T109_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T120_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T129_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T134_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T150_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T153_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T156_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T15_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T26_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T33_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T47_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T73_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oce

/home/hmiwa/hmiwa1/rawdata/Oceania/t112_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t138_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t159_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t160_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t21_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t38_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t51_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t69_sorted.genome.vcf.gz


In [78]:
#準備
bcftools view --header-only $rawdir/T03_sorted.genome.vcf.gz | tail -n 1
bcftools view --no-header $rawdir/T03_sorted.genome.vcf.gz | head -n 1

#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	T03
chrM	1	.	G	.	.	LowGQX	.	GT:GQX:DP:DPF:MIN_DP	0/0:12:5:0:5


In [88]:
#準備（ツールの変数設定）
vcftools=/home/hmiwa/homebrew/bin/vcftools

データフォルダから読み出して、FILTER=PASSのものをGTだけにする

In [96]:
cd $strdir/targetdata
for t in T03 T104 T105 T106 T109 T120 T129 T134 T150 T153 T156 T15 T26 T33 T47 T73 T74 T76 T82 T85 T87 T88 T98 t112 t138 t159 t160 t21 t38 t51 t69
do
$vcftools --gzvcf $rawdir/${t}_sorted.genome.vcf.gz --remove-filtered-all --recode --stdout | bgzip -c > ${t}_sorted.genome_PASS_only.vcf.gz
bcftools index ${t}_sorted.genome_PASS_only.vcf.gz
for i in {1..22}
do
bcftools view --header-only  ${t}_sorted.genome_PASS_only.vcf.gz > ${t}_sorted.genome_PASS_only.chr${i}.vcf
bcftools query -r chr$i -f '%CHROM\t%POS\t%ID\t%REF\t%ALT\t%QUAL\t%FILTER\t%INFO\tGT[\t%GT]\n' ${t}_sorted.genome_PASS_only.vcf.gz | sed s/^chr//g >> ${t}_sorted.genome_PASS_only.chr${i}.vcf
bgzip -c ${t}_sorted.genome_PASS_only.chr${i}.vcf > ${t}_sorted.genome_PASS_only.chr${i}.vcf.gz
rm ${t}_sorted.genome_PASS_only.chr${i}.vcf
bcftools index ${t}_sorted.genome_PASS_only.chr${i}.vcf.gz
done
done

トンガ人サンプルをマージしたファイルを作る

In [97]:
cd $strdir/targetdata
for i in {1..22}
do
f1=`for t in T03 T104 T105 T106 T109 T120 T129 T134 T150 T153 T156 T15 T26 T33 T47 T73 T74 T76 T82 T85 T87 T88 T98 t112 t138 t159 t160 t21 t38 t51 t69; do echo "${t}_sorted.genome_PASS_only.chr${i}.vcf.gz "; done`
f=`echo $f1 | awk '{printf $0}'`
`echo "bcftools merge --missing-to-ref -r ${i} -Oz -o Tongan.chr${i}.vcf.gz ${f}"`
bcftools index Tongan.chr${i}.vcf.gz
done

[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types

マニュアルで推奨されているplinkコマンドは以下の通り

    plink \
        --bfile EUR \
        --maf 0.01 \
        --hwe 1e-6 \
        --geno 0.01 \
        --mind 0.01 \
        --write-snplist \
        --make-just-fam \
        --out EUR.QC

Beagleなどで処理する必要があるのでMAFやHWEなどはこの段階である程度フィルタリングしておく


Beagle処理の前にfixploidyをしておく必要がある

In [101]:
bcftools=/home/hmiwa/homebrew/bin/bcftools

In [102]:
cd $strdir/targetdata
for i in {1..22}
do
$vcftools --gzvcf Tongan.chr${i}.vcf.gz \
--hwe 0.000001  --max-missing 0.99 --maf 0.01 \
--remove-indels --min-alleles 2 --max-alleles 2 \
--remove-filtered-all --recode --stdout | bgzip -c > Tongan.chr${i}.QC.vcf.gz
$bcftools index Tongan.chr${i}.QC.vcf.gz
$bcftools +fixploidy Tongan.chr${i}.QC.vcf.gz -Oz --threads 24 -o Tongan.chr${i}.QC.FIXP.vcf.gz -- --default-ploidy 2
$bcftools index Tongan.chr${i}.QC.FIXP.vcf.gz
done

Beagle関係のディレクトリを変数に入れておく

In [103]:
beagle=/home/hmiwa/local/beagle.22Jul22.46e.jar
beagle_map=/home/hmiwa/local/plink.GRCh37.map

Beagleでフェージング処理をしておく

In [104]:
cd $strdir/targetdata
for i in {1..22}
do
java -Xmx16g -jar $beagle \
gt=Tongan.chr${i}.QC.FIXP.vcf.gz \
out=Tongan.chr${i}.QC.FIXP.phased \
map=${beagle_map}/plink.chr${i}.GRCh37.map \
nthreads=16 | tee Tongan.chr${i}.QC.FIXP.phased.log
done

beagle.22Jul22.46e.jar (version 5.4)
Copyright (C) 2014-2022 Brian L. Browning
Enter "java -jar beagle.22Jul22.46e.jar" to list command line argument
Start time: 10:55 PM JST on 18 Dec 2023

Command line: java -Xmx16384m -jar beagle.22Jul22.46e.jar
  gt=Tongan.chr1.QC.FIXP.vcf.gz
  out=Tongan.chr1.QC.FIXP.phased
  map=/home/hmiwa/local/plink.GRCh37.map/plink.chr1.GRCh37.map
  nthreads=16

Reference samples:                    0
Study     samples:                   31

Window 1 [1:10450-18870356]
Study     markers:               71,192

Burnin  iteration 1:           3 seconds
Burnin  iteration 2:           1 second
Burnin  iteration 3:           1 second

Estimated ne:                  24080
Estimated err:                 8.9e-03

Phasing iteration 1:           1 second
Phasing iteration 2:           0 seconds
Phasing iteration 3:           0 seconds
Phasing iteration 4:           0 seconds
Phasing iteration 5:           0 seconds
Phasing iteration 6:           0 seconds
Phasing iterat

そのまえにこの段階でSNP IDを付与しておかねばならない。まず前段階の準備を行う。

In [105]:
cd $strdir/targetdata
#SNPID付与の準備（中身はVCF形式なので拡張子の変更をする）
gunzip -c ~/hmiwa1/rawdata/SNP/GCF_000001405.25.gz | bgzip -c > GCF_000001405.25.vcf.gz
bcftools index GCF_000001405.25.vcf.gz

In [106]:
#上記はCHROMがalias sequence namesになっているので変更する
wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.chromAlias.txt
wget https://raw.githubusercontent.com/humgendiv/hmiwa-shared/main/get_rsID/aliasrename.R
Rscript aliasrename.R

echo "start annotating"
bcftools annotate --rename-chrs alias.new.ids \
-Oz -o GCF_000001405.25.annotated.vcf.gz GCF_000001405.25.vcf.gz
echo "... annotated"
echo "start indexing"
bcftools index GCF_000001405.25.annotated.vcf.gz
echo "... indexed"

--2023-12-19 00:57:13--  http://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.chromAlias.txt
Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163
Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 16866 (16K) [text/plain]
Saving to: ‘hg19.chromAlias.txt’

hg19.chromAlias.txt 100%[===================>]  16.47K  --.-KB/s    in 0.1s    

2023-12-19 00:57:14 (154 KB/s) - ‘hg19.chromAlias.txt’ saved [16866/16866]

--2023-12-19 00:57:14--  https://raw.githubusercontent.com/humgendiv/hmiwa-shared/main/get_rsID/aliasrename.R
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 402 [text/plain]
Saving to: ‘aliasrename.R’

aliasrename.R    

In [107]:
#染色体ごとに分割する
echo "start separating by chromosomes"
for i in {1..22}
do
bcftools view -r chr${i} GCF_000001405.25.annotated.vcf.gz | \
sed -e s/^chr//g | bgzip -c > GCF_000001405.25.annotated.chr${i}.vcf.gz
echo "... chr${i} separated"
bcftools index GCF_000001405.25.annotated.chr${i}.vcf.gz
echo "... chr${i} indexed"
done

start separating by chromosomes
... chr1 separated
... chr1 indexed
... chr2 separated
... chr2 indexed
... chr3 separated
... chr3 indexed
... chr4 separated
... chr4 indexed
... chr5 separated
... chr5 indexed
... chr6 separated
... chr6 indexed
... chr7 separated
... chr7 indexed
... chr8 separated
... chr8 indexed
... chr9 separated
... chr9 indexed
... chr10 separated
... chr10 indexed
... chr11 separated
... chr11 indexed
... chr12 separated
... chr12 indexed
... chr13 separated
... chr13 indexed
... chr14 separated
... chr14 indexed
... chr15 separated
... chr15 indexed
... chr16 separated
... chr16 indexed
... chr17 separated
... chr17 indexed
... chr18 separated
... chr18 indexed
... chr19 separated
... chr19 indexed
... chr20 separated
... chr20 indexed
... chr21 separated
... chr21 indexed
... chr22 separated
... chr22 indexed


アノテーションを実行する。

In [1]:
ls

05_231211_RePRSTrials.ipynb  Documents  jupyter_venv5  pic  test.ipynb


In [5]:
cd $strdir/targetdata

echo START: `date`
#常染色体すべてにrsIDを付与する場合
for i in {1..22}
do
INFILE=Tongan.chr${i}.QC.FIXP.phased.vcf.gz
REFFILE=GCF_000001405.25.annotated.chr${i}.vcf.gz
OUTFILE=Tongan.chr${i}.QC.FIXP.phased.annotated.vcf.gz
bcftools index $INFILE
bcftools annotate -a $REFFILE -Oz -o $OUTFILE -c ID $INFILE --threads 16
bcftools index $OUTFILE
done
echo END: `date`

START: 2023年 12月 19日 火曜日 17:08:39 JST
[E::main_vcfindex] the index file exists. Please use '-f' to overwrite Tongan.chr22.QC.FIXP.phased.vcf.gz.csi
END: 2023年 12月 19日 火曜日 17:25:20 JST


常染色体をひとつのファイルにまとめる

In [7]:
bcftools concat \
    -Oz -o Tongan.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr1.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr2.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr3.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr4.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr5.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr6.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr7.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr8.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr9.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr10.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr11.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr12.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr13.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr14.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr15.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr16.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr17.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr18.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr19.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr20.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr21.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr22.QC.FIXP.phased.annotated.vcf.gz
bcftools index Tongan.QC.FIXP.phased.annotated.vcf.gz

Checking the headers and starting positions of 22 files
Concatenating Tongan.chr1.QC.FIXP.phased.annotated.vcf.gz	2.630552 seconds
Concatenating Tongan.chr2.QC.FIXP.phased.annotated.vcf.gz	2.533487 seconds
Concatenating Tongan.chr3.QC.FIXP.phased.annotated.vcf.gz	2.074929 seconds
Concatenating Tongan.chr4.QC.FIXP.phased.annotated.vcf.gz	2.101666 seconds
Concatenating Tongan.chr5.QC.FIXP.phased.annotated.vcf.gz	1.768070 seconds
Concatenating Tongan.chr6.QC.FIXP.phased.annotated.vcf.gz	1.859544 seconds
Concatenating Tongan.chr7.QC.FIXP.phased.annotated.vcf.gz	1.723798 seconds
Concatenating Tongan.chr8.QC.FIXP.phased.annotated.vcf.gz	1.509193 seconds
Concatenating Tongan.chr9.QC.FIXP.phased.annotated.vcf.gz	1.394220 seconds
Concatenating Tongan.chr10.QC.FIXP.phased.annotated.vcf.gz	1.418890 seconds
Concatenating Tongan.chr11.QC.FIXP.phased.annotated.vcf.gz	1.395065 seconds
Concatenating Tongan.chr12.QC.FIXP.phased.annotated.vcf.gz	1.363414 seconds
Concatenating Tongan.chr13.QC.FIXP.phased

ここからはPLINK形式のファイルで作業したいので変換しておく。

In [8]:
plink \
    --vcf Tongan.QC.FIXP.phased.annotated.vcf.gz \
    --make-bed \
    --out TNG

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.log.
Options in effect:
  --make-bed
  --out TNG
  --vcf Tongan.QC.FIXP.phased.annotated.vcf.gz

257564 MB RAM detected; reserving 128782 MB for main workspace.
--vcf: TNG-temporary.bed + TNG-temporary.bim + TNG-temporary.fam written.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nosex .
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
8711185 variants and 31 people pass filters and QC.
Note: No phenoty

In [9]:
ls TNG*

TNG.bed  TNG.bim  TNG.fam  TNG.log  TNG.nosex


ジェノタイピング率が低いSNP、マイナーアレル頻度が低いSNP、ハーディーワインベルグ平衡から外れたSNP、ジェノタイピング率が低い個体などの除去を行う。

In [10]:
plink \
    --bfile TNG \
    --maf 0.01 \
    --hwe 1e-6 \
    --geno 0.01 \
    --mind 0.01 \
    --write-snplist \
    --make-just-fam \
    --out TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --geno 0.01
  --hwe 1e-6
  --maf 0.01
  --make-just-fam
  --mind 0.01
  --out TNG.QC
  --write-snplist

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
0 people removed due to missing genotype data (--mind).
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
0 variants removed due to missing

In [11]:
ls TNG*

TNG.bed  TNG.fam  TNG.nosex   TNG.QC.log    TNG.QC.snplist
TNG.bim  TNG.log  TNG.QC.fam  TNG.QC.nosex


極端なヘテロ接合度を持つ個体を除去する。

まずは相関性の高いSNPを除去するためにプルーニング

In [12]:
plink \
    --bfile TNG \
    --keep TNG.QC.fam \
    --extract TNG.QC.snplist \
    --indep-pairwise 200 50 0.25 \
    --out TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --extract TNG.QC.snplist
  --indep-pairwise 200 50 0.25
  --keep TNG.QC.fam
  --out TNG.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
--extract: 8708432 variants remaining.
--keep: 31 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
8708432 variants and 31 people pa

ヘテロ接合度を計算する

In [13]:
plink \
    --bfile TNG \
    --extract TNG.QC.prune.in \
    --keep TNG.QC.fam \
    --het \
    --out TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --extract TNG.QC.prune.in
  --het
  --keep TNG.QC.fam
  --out TNG.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
--extract: 660791 variants remaining.
--keep: 31 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
660791 variants and 31 people pass filters and QC.
Note:

現在のディレクトリチェック

In [4]:
cd $strdir
ls

Basedata  StatGenSummerSchool2023_GenomeDataAnalysis5      targetdata
out.log   StatGenSummerSchool2023_GenomeDataAnalysis5.zip


ところでVScodeでjupyterのカーネルが選べないときはjupyterの拡張機能のバージョンを落としてみてくださいね（蛇足）

In [134]:
cd $strdir
cat << EOL > $gitdir/het.R
setwd("$strdir/targetdata")
dat <- read.table("TNG.QC.het", header=T) # Read in the EUR.het file, specify it has header
m <- mean(dat\$F) # Calculate the mean
s <- sd(dat\$F) # Calculate the SD
valid <- subset(dat, F <= m+3*s & F >= m-3*s) # Get any samples with F coefficient within 3 SD of the population mean
write.table(valid[,c(1,2)], "TNG.valid.sample", quote=F, row.names=F) # print FID and IID for valid samples
EOL
cat $gitdir/het.R

setwd("/home/hmiwa/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials/targetdata")
dat <- read.table("TNG.QC.het", header=T) # Read in the EUR.het file, specify it has header
m <- mean(dat$F) # Calculate the mean
s <- sd(dat$F) # Calculate the SD
valid <- subset(dat, F <= m+3*s & F >= m-3*s) # Get any samples with F coefficient within 3 SD of the population mean
write.table(valid[,c(1,2)], "TNG.valid.sample", quote=F, row.names=F) # print FID and IID for valid samples


In [6]:
Rscript --no-save $gitdir/het.R

In [8]:
cd $strdir/targetdata
ls TNG*

TNG.bed  TNG.log     TNG.QC.het    TNG.QC.prune.in   TNG.valid.sample
TNG.bim  TNG.nosex   TNG.QC.log    TNG.QC.prune.out
TNG.fam  TNG.QC.fam  TNG.QC.nosex  TNG.QC.snplist


In [9]:
head TNG.valid.sample

FID IID
T03 T03
T104 T104
T105 T105
T106 T106
T109 T109
T120 T120
T129 T129
T134 T134
T150 T150


In [10]:
wc TNG.valid.sample
wc TNG.QC.het

 32  64 284 TNG.valid.sample
  32  192 1984 TNG.QC.het


In [128]:
head TNG.QC.het

 FID  IID       O(HOM)       E(HOM)        N(NM)            F
 T03  T03       556473    5.574e+05       660791    -0.008687
T104 T104       560906    5.574e+05       660791      0.03418
T105 T105       562118    5.574e+05       660791       0.0459
T106 T106       560067    5.574e+05       660791      0.02606
T109 T109       560892    5.574e+05       660791      0.03404
T120 T120       563083    5.574e+05       660791      0.05523
T129 T129       558374    5.574e+05       660791     0.009695
T134 T134       557317    5.574e+05       660791   -0.0005258
T150 T150       561071    5.574e+05       660791      0.03577


In [145]:
wc TNG.valid.sample
wc TNG.fam

 32  64 284 TNG.valid.sample


 31 186 555 TNG.fam


##### Ambiguous SNPs

ここはBaseDataで削除済みなのでOK

##### Mismatching SNPs

ほとんどのPRSソフトウェアは自動的にチェックするらしいので不要。

もしうまくいかなかったら考えればよさそう。

##### Duplicate SNPs

ターゲットデータ中の重複SNPは削除しろとのこと。

Beagleが動作している時点で重複SNPは削除されているはずなので、問題はなかろう。

##### Sex Chromosomes

上記ファイルはX染色体を含まないので、念の為別途チェックする。

In [5]:
#準備（ツールの変数設定）
vcftools=/home/hmiwa/homebrew/bin/vcftools
bcftools=/home/hmiwa/homebrew/bin/bcftools

In [15]:
#準備（directoryの変数設定）
strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
rawdir=~/hmiwa1/rawdata/Oceania

In [17]:
cd $strdir/targetdata
#FILTER=PASSだけ拾う
for t in T03 T104 T105 T106 T109 T120 T129 T134 T150 T153 T156 T15 T26 T33 T47 T73 T74 T76 T82 T85 T87 T88 T98 t112 t138 t159 t160 t21 t38 t51 t69
do
i="X"
bcftools view --header-only  ${t}_sorted.genome_PASS_only.vcf.gz > ${t}_sorted.genome_PASS_only.chr${i}.vcf
bcftools query -r chr$i -f '%CHROM\t%POS\t%ID\t%REF\t%ALT\t%QUAL\t%FILTER\t%INFO\tGT[\t%GT]\n' ${t}_sorted.genome_PASS_only.vcf.gz | sed s/^chr//g >> ${t}_sorted.genome_PASS_only.chr${i}.vcf
bgzip -c ${t}_sorted.genome_PASS_only.chr${i}.vcf > ${t}_sorted.genome_PASS_only.chr${i}.vcf.gz
rm ${t}_sorted.genome_PASS_only.chr${i}.vcf
bcftools index ${t}_sorted.genome_PASS_only.chr${i}.vcf.gz
done

In [18]:
cd $strdir/targetdata
#サンプルのマージ
f1=`for t in T03 T104 T105 T106 T109 T120 T129 T134 T150 T153 T156 T15 T26 T33 T47 T73 T74 T76 T82 T85 T87 T88 T98 t112 t138 t159 t160 t21 t38 t51 t69; do echo "${t}_sorted.genome_PASS_only.chr${i}.vcf.gz "; done`
f=`echo $f1 | awk '{printf $0}'`
`echo "bcftools merge --missing-to-ref -r ${i} -Oz -o Tongan.chr${i}.vcf.gz ${f}"`
bcftools index Tongan.chr${i}.vcf.gz

[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types


In [19]:
cd $strdir/targetdata
#phasing前処理
$vcftools --gzvcf Tongan.chr${i}.vcf.gz \
--hwe 0.000001  --max-missing 0.99 --maf 0.01 \
--remove-indels --min-alleles 2 --max-alleles 2 \
--remove-filtered-all --recode --stdout | bgzip -c > Tongan.chr${i}.QC.vcf.gz
$bcftools index Tongan.chr${i}.QC.vcf.gz
$bcftools +fixploidy Tongan.chr${i}.QC.vcf.gz -Oz --threads 24 -o Tongan.chr${i}.QC.FIXP.vcf.gz -- --default-ploidy 2
$bcftools index Tongan.chr${i}.QC.FIXP.vcf.gz

#beagle変数
beagle=/home/hmiwa/local/beagle.22Jul22.46e.jar
beagle_map=/home/hmiwa/local/plink.GRCh37.map

In [40]:
#beagleがうまくいかなかったのでチェック
cd $strdir/targetdata
$bcftools view --no-header Tongan.chrX.QC.FIXP.vcf.gz | head -n 1
$bcftools view --no-header Tongan.chrX.QC.FIXP.vcf.gz | grep 2832470
$bcftools view --no-header Tongan.chrX.QC.FIXP.vcf.gz | wc
$bcftools view --no-header Tongan.chrX.QC.FIXP.vcf.gz | grep -e $'\t'1$'\t' 
$bcftools view --no-header Tongan.chrX.QC.FIXP.vcf.gz | grep -e $'\t'0$'\t' | wc
$bcftools view --no-header Tongan.chrX.QC.FIXP.vcf.gz | grep -v $'\t'0$'\t' | grep -v $'\t'1$'\t' | wc

X	60015	.	A	G	52	PASS	.	GT	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/1
X	2832470	.	G	C	169	PASS	.	GT	0	0	0	0	0	0	1	0	0	0	0	0	0	1	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0
 149807 5992280 23212450
X	230716	.	G	A	1	PASS	.	GT	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0|1	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0|1	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0
X	386205	.	A	G	1	PASS	.	GT	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/1	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0
X	523419	.	A	G	1	PASS	.	GT	0/0	0/0	0/0	0/1	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/1	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0
X	623749	.	T	A	1	PASS	.	GT	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/1	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0
X	813467	.	C	G	1	PASS	.	GT	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/0	0/1	0/0	0/0	0/0	0|1	0/0	0/0	0/0	0/0	0|1	0/0	0/0	0

In [41]:
bcftools view --header-only Tongan.chrX.QC.FIXP.vcf.gz > Tongan.chrX.QC.FIXP.2.vcf
bcftools view --no-header Tongan.chrX.QC.FIXP.vcf.gz | grep -v $'\t'0$'\t' >> Tongan.chrX.QC.FIXP.2.vcf
bgzip -c Tongan.chrX.QC.FIXP.2.vcf > Tongan.chrX.QC.FIXP.2.vcf.gz
rm Tongan.chrX.QC.FIXP.2.vcf
bcftools index Tongan.chrX.QC.FIXP.2.vcf.gz

In [42]:
#beagle(phasing)
i="X"
java -Xmx16g -jar $beagle \
gt=Tongan.chr${i}.QC.FIXP.2.vcf.gz \
out=Tongan.chr${i}.QC.FIXP.phased \
map=${beagle_map}/plink.chr${i}.GRCh37.map \
nthreads=16 | tee Tongan.chr${i}.QC.FIXP.phased.log

beagle.22Jul22.46e.jar (version 5.4)
Copyright (C) 2014-2022 Brian L. Browning
Enter "java -jar beagle.22Jul22.46e.jar" to list command line argument
Start time: 12:56 AM JST on 24 Dec 2023

Command line: java -Xmx16384m -jar beagle.22Jul22.46e.jar
  gt=Tongan.chrX.QC.FIXP.2.vcf.gz
  out=Tongan.chrX.QC.FIXP.phased
  map=/home/hmiwa/local/plink.GRCh37.map/plink.chrX.GRCh37.map
  nthreads=16

Reference samples:                    0
Study     samples:                   31

Window 1 [X:60015-23130231]
Study     markers:               37,834

Burnin  iteration 1:           3 seconds
Burnin  iteration 2:           0 seconds
Burnin  iteration 3:           0 seconds

Estimated ne:                  22600
Estimated err:                 4.5e-02

Phasing iteration 1:           0 seconds
Phasing iteration 2:           0 seconds
Phasing iteration 3:           0 seconds
Phasing iteration 4:           0 seconds
Phasing iteration 5:           0 seconds
Phasing iteration 6:           0 seconds
Phasing i

In [43]:
#annotation準備
echo "start separating by chromosomes"
bcftools view -r chr${i} GCF_000001405.25.annotated.vcf.gz | \
sed -e s/^chr//g | bgzip -c > GCF_000001405.25.annotated.chr${i}.vcf.gz
echo "... chr${i} separated"
bcftools index GCF_000001405.25.annotated.chr${i}.vcf.gz
echo "... chr${i} indexed"

start separating by chromosomes
... chrX separated
... chrX indexed


In [44]:
#annotation
cd $strdir/targetdata
echo START: `date`
INFILE=Tongan.chr${i}.QC.FIXP.phased.vcf.gz
REFFILE=GCF_000001405.25.annotated.chr${i}.vcf.gz
OUTFILE=Tongan.chr${i}.QC.FIXP.phased.annotated.vcf.gz
bcftools index $INFILE
bcftools annotate -a $REFFILE -Oz -o $OUTFILE -c ID $INFILE --threads 16
bcftools index $OUTFILE
echo END: `date`

START: 2023年 12月 24日 日曜日 01:02:55 JST
END: 2023年 12月 24日 日曜日 01:03:35 JST


In [45]:
plink \
    --vcf Tongan.chrX.QC.FIXP.phased.annotated.vcf.gz \
    --make-bed \
    --out TNG.X

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.X.log.
Options in effect:
  --make-bed
  --out TNG.X
  --vcf Tongan.chrX.QC.FIXP.phased.annotated.vcf.gz

257564 MB RAM detected; reserving 128782 MB for main workspace.
--vcf: TNG.X-temporary.bed + TNG.X-temporary.bim + TNG.X-temporary.fam written.
147678 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.X.nosex .
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
147678 variants and 31 people pass filters and QC.
N

In [46]:
plink \
    --bfile TNG.X \
    --maf 0.01 \
    --hwe 1e-6 \
    --geno 0.01 \
    --mind 0.01 \
    --write-snplist \
    --make-just-fam \
    --out TNG.X.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.X.QC.log.
Options in effect:
  --bfile TNG.X
  --geno 0.01
  --hwe 1e-6
  --maf 0.01
  --make-just-fam
  --mind 0.01
  --out TNG.X.QC
  --write-snplist

257564 MB RAM detected; reserving 128782 MB for main workspace.
147678 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.X.QC.nosex .
0 people removed due to missing genotype data (--mind).
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
0 variants removed due to 

In [47]:
plink \
    --bfile TNG.X \
    --keep TNG.X.QC.fam \
    --extract TNG.X.QC.snplist \
    --indep-pairwise 200 50 0.25 \
    --out TNG.X.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.X.QC.log.
Options in effect:
  --bfile TNG.X
  --extract TNG.X.QC.snplist
  --indep-pairwise 200 50 0.25
  --keep TNG.X.QC.fam
  --out TNG.X.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
147678 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.X.QC.nosex .
--extract: 147657 variants remaining.
--keep: 31 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
147657 variants and 31 

In [48]:
plink \
    --bfile TNG.X \
    --extract TNG.X.QC.prune.in \
    --keep TNG.valid.sample \
    --check-sex \
    --out TNG.X.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.X.QC.log.
Options in effect:
  --bfile TNG.X
  --check-sex
  --extract TNG.X.QC.prune.in
  --keep TNG.valid.sample
  --out TNG.X.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
147678 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.X.QC.nosex .
--extract: 16384 variants remaining.
--keep: 31 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
16384 variants and 31 people pass fi

In [49]:
cat TNG.X.QC.sexcheck

 FID  IID       PEDSEX       SNPSEX       STATUS            F
 T03  T03            0            0      PROBLEM       0.6585
T104 T104            0            0      PROBLEM       0.6546
T105 T105            0            0      PROBLEM       0.6301
T106 T106            0            0      PROBLEM       0.6143
T109 T109            0            0      PROBLEM       0.6311
T120 T120            0            0      PROBLEM       0.6746
T129 T129            0            0      PROBLEM       0.6385
T134 T134            0            0      PROBLEM       0.6189
T150 T150            0            0      PROBLEM       0.6592
T153 T153            0            0      PROBLEM       0.6903
T156 T156            0            0      PROBLEM       0.6875
 T15  T15            0            0      PROBLEM        0.677
 T26  T26            0            0      PROBLEM       0.6819
 T33  T33            0            0      PROBLEM       0.6672
 T47  T47            0            0      PROBLEM       0.6609
 T73  T7

In [50]:
column -s , -t $gitdir/Oceania_Tonga_231223.csv

IDNO  AGE  SEX  Height  Weight  BMI
T3    53   0    167.5   73.2    26.09044331
T104  50   0    173.4   107.9   35.88585173
T105  55   0    179.1   131.8   41.08887386
T106  55   0    172.2   109.3   36.85987582
T109  21   0    171.5   78      26.51956243
T120  47   0    175.9   92      29.73419248
T129  51   0    177.5   102     32.37452886
T134  36   0    177.1   100.4   32.0108046
T150  42   0    170.3   104.9   36.16980707
T153  60   0    176.2   168.5   54.27353346
T156  43   0    172     98.4    33.26122228
T15   57   0    172.8   93.1    31.17900163
T26   53   0    178.3   130.7   41.11240673
T33   58   0    173.5   114.3   37.97058359
T47   32   0    184.2   121.8   35.89781678
T73   40   0    168.5   60.5    21.30863176
T74   50   0    167.4   89.6    31.9740097
T76   63   0    166.2   106.6   38.59181159
T82   24   0    181.5   79.2    24.04207363
T85   42   0    172.2   96.9    32.67815157
T87   33   0    170.5   94.2    32.40426209
T88   47   0    178.7   143     44.7802931

すべてのサンプルが男性とラベリングされているので、おそらく齟齬はないだろう

##### Sample Overlap

ターゲットデータとベースデータで重複するものはない

##### Relatedness

In [53]:
cd $strdir/targetdata
plink \
    --bfile TNG \
    --extract TNG.QC.prune.in \
    --keep TNG.valid.sample \
    --rel-cutoff 0.125 \
    --out TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --extract TNG.QC.prune.in
  --keep TNG.valid.sample
  --out TNG.QC
  --rel-cutoff 0.125

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
--extract: 660791 variants remaining.
--keep: 31 people remaining.
Using up to 63 threads (change this with --threads).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
660791 variants and 31 people pass filte

In [54]:
head TNG.QC.rel.id
wc TNG.QC.rel.id

T03	T03
T104	T104
T105	T105
T106	T106
T109	T109
T120	T120
T129	T129
T134	T134
T150	T150
T153	T153
 31  62 276 TNG.QC.rel.id


PLINKのRelatednessではどのサンプルも除去されなかったらしい。(pruning あり)

以前のRelatedness解析ではT120とT156、T73とT87が近縁と出ている（過去ログ参照）

In [57]:
cd $strdir/targetdata
plink \
    --bfile TNG \
    --keep TNG.valid.sample \
    --rel-cutoff 0.125 \
    --out TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --keep TNG.valid.sample
  --out TNG.QC
  --rel-cutoff 0.125

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
--keep: 31 people remaining.
Using up to 63 threads (change this with --threads).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
8711185 variants and 31 people pass filters and QC (before --rel-cutoff).
Note: No phenotypes present.
Rel

In [59]:
cat TNG.QC.rel.id
wc TNG.QC.rel.id

T03	T03
T104	T104
T105	T105
T106	T106
T109	T109
T120	T120
T129	T129
T134	T134
T150	T150
T153	T153
T15	T15
T26	T26
T33	T33
T47	T47
T73	T73
T74	T74
T76	T76
T82	T82
T85	T85
T87	T87
T88	T88
T98	T98
t112	t112
t138	t138
t159	t159
t160	t160
t21	t21
t38	t38
t51	t51
t69	t69


 30  60 266 TNG.QC.rel.id


このバージョンではT156が除去されている（pruning なし）

一旦すべてのサンプルを含めておく。

In [56]:
ls TNG*

TNG.bed     TNG.QC.log        TNG.X.bed     TNG.X.QC.nosex
TNG.bim     TNG.QC.nosex      TNG.X.bim     TNG.X.QC.prune.in
TNG.fam     TNG.QC.prune.in   TNG.X.fam     TNG.X.QC.prune.out
TNG.log     TNG.QC.prune.out  TNG.X.log     TNG.X.QC.sexcheck
TNG.nosex   TNG.QC.rel.id     TNG.X.nosex   TNG.X.QC.snplist
TNG.QC.fam  TNG.QC.snplist    TNG.X.QC.fam
TNG.QC.het  TNG.valid.sample  TNG.X.QC.log


##### Generate final QC'ed target data file

In [60]:
plink \
    --bfile TNG \
    --make-bed \
    --keep TNG.QC.fam \
    --extract TNG.QC.snplist \
    --out TNG.QC
    #とりあえずすべてのサンプルを含める 

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --extract TNG.QC.snplist
  --keep TNG.QC.fam
  --make-bed
  --out TNG.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
--extract: 8708432 variants remaining.
--keep: 31 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
8708432 variants and 31 people pass filters and QC.

### Calculating and Analysing PRS

In [3]:
#準備（directoryの変数設定）
strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
rawdir=~/hmiwa1/rawdata/Oceania

In [4]:
cd $strdir
ls

Basedata  StatGenSummerSchool2023_GenomeDataAnalysis5      targetdata
out.log   StatGenSummerSchool2023_GenomeDataAnalysis5.zip


#### Basedataのチェック

In [5]:
ls Basedata

BMI_Eastern_Asian.fmt.gzip    BMI_Eastern_Asian.QC.txt
BMI_Eastern_Asian.newcol.txt  BMI_Eastern_Asian.txt
BMI_Eastern_Asian.nodup.txt


In [6]:
Basedata/BMI_Eastern_Asian.QC.txthead 

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100017453	T	G	rs1983864	G	0.6081	G	0.5682	-	-0.031	0.015	0.044
10	100017532	G	A	rs77858766	A	0.0169	A	0.01781	-	-0.016	0.058	0.79
10	100017553	G	T	rs11189525	T	0.0129	T	0.007404	-	-0.014	0.13	0.91
10	100020880	C	T	rs33995374	T	0.1062	T	0.1163	0.077	0.061	0.21
10	100044136	G	A	rs4919199	A	0.9544	-	-	0.0076	0.036	0.83
10	100144782	C	T	rs2296441	T	0.4425	T	0.42	0.012	0.015	0.42
10	100148004	T	C	rs2182169	C	0.5615	-	-	0.014	0.015	0.37
10	100148176	A	G	rs2147896	G	0.5833	G	0.5596	0.0052	0.015	0.73


BetaでBaseDataが用意されているのでORから変更する必要はなさそう。

~~項目の順番やヘッダーを直す必要はある。~~

In [13]:
head Basedata/BMI_Eastern_Asian.QC.txt | awk '{print $1,$2,$5,$3,$4,"8839",$11,$12,$7,$10}'

CHR POS SNPNAME REF ALT 8839 se Pvalue EAS_MAF beta
10 100010909 rs141069497 C T 8839 -0.022 0.1 0.0139 -
10 100017453 rs1983864 T G 8839 -0.031 0.015 0.6081 -
10 100017532 rs77858766 G A 8839 -0.016 0.058 0.0169 -
10 100017553 rs11189525 G T 8839 -0.014 0.13 0.0129 -
10 100020880 rs33995374 C T 8839 0.061 0.21 0.1062 0.077
10 100044136 rs4919199 G A 8839 0.036 0.83 0.9544 0.0076
10 100144782 rs2296441 C T 8839 0.015 0.42 0.4425 0.012
10 100148004 rs2182169 T C 8839 0.015 0.37 0.5615 0.014
10 100148176 rs2147896 A G 8839 0.015 0.73 0.5833 0.0052


#### 追加QC

トンガ人SNPリストに”.”が入っていたので除去します。


In [26]:
ls targetdata/TNG.QC*

targetdata/TNG.QC.bed  targetdata/TNG.QC.log        targetdata/TNG.QC.rel.id
targetdata/TNG.QC.bim  targetdata/TNG.QC.nosex      targetdata/TNG.QC.snplist
targetdata/TNG.QC.fam  targetdata/TNG.QC.prune.in
targetdata/TNG.QC.het  targetdata/TNG.QC.prune.out


In [40]:
cat targetdata/TNG.QC.snplist | wc
cat targetdata/TNG.QC.snplist | grep rs | wc
cat targetdata/TNG.QC.snplist | grep -v rs | wc
cat targetdata/TNG.QC.snplist | grep -v rs | head -n 5

8708431 8708431 93962806
8612836 8612836 93771616
  95595   95595  191190
.
.
.
.
.


In [44]:
cat targetdata/TNG.QC.snplist | wc
cat targetdata/TNG.QC.snplist | grep rs | wc
cat targetdata/TNG.QC.snplist | grep rs > targetdata/TNG.QC.snplist.rs

8708431 8708431 93962806
8612836 8612836 93771616


Pvalueに負の値や１があるらしいので除去します。

In [46]:
head -n 2 Basedata/BMI_Eastern_Asian.QC.txt

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83


In [54]:
awk '$12 < 0.0' Basedata/BMI_Eastern_Asian.QC.txt | wc
awk '$12 > 0.0' Basedata/BMI_Eastern_Asian.QC.txt | wc
awk '$12 > 1.0' Basedata/BMI_Eastern_Asian.QC.txt | wc
awk '$12 < 1.0 && $12 > 0.0' Basedata/BMI_Eastern_Asian.QC.txt | wc

     59     826    3662
  28460  355430 1704171
    528    6339   24940
  27872  348334 1676024


In [65]:
head -n 1 Basedata/BMI_Eastern_Asian.QC.txt > Basedata/BMI_Eastern_Asian.QC2.txt
awk '$12 < 1.0 && $12 > 0.0' Basedata/BMI_Eastern_Asian.QC.txt >> Basedata/BMI_Eastern_Asian.QC2.txt

In [66]:
wc Basedata/BMI_Eastern_Asian.QC.txt
wc Basedata/BMI_Eastern_Asian.QC2.txt

  28519  356256 1707833 Basedata/BMI_Eastern_Asian.QC.txt


  27873  348346 1676097 Basedata/BMI_Eastern_Asian.QC2.txt


#### Clumping

In [67]:
cd $strdir
plink \
    --bfile targetdata/TNG.QC \
    --extract targetdata/TNG.QC.snplist.rs \
    --clump-p1 1 \
    --clump-r2 0.1 \
    --clump-kb 250 \
    --clump Basedata/BMI_Eastern_Asian.QC2.txt \
    --clump-snp-field SNPNAME \
    --clump-field Pvalue \
    --out TNG

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.log.
Options in effect:
  --bfile targetdata/TNG.QC
  --clump Basedata/BMI_Eastern_Asian.QC2.txt
  --clump-field Pvalue
  --clump-kb 250
  --clump-p1 1
  --clump-r2 0.1
  --clump-snp-field SNPNAME
  --extract targetdata/TNG.QC.snplist.rs
  --out TNG

257564 MB RAM detected; reserving 128782 MB for main workspace.
8708432 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nosex .
--extract: 8612836 variants remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949

In [68]:
awk 'NR!=1{print $3}' TNG.clumped >  TNG.valid.snp

In [69]:
ls

Basedata                                         TNG.clumped
out.log                                          TNG.log
StatGenSummerSchool2023_GenomeDataAnalysis5      TNG.nosex
StatGenSummerSchool2023_GenomeDataAnalysis5.zip  TNG.valid.snp
targetdata


### PRSice-2を動かしてみよう

In [70]:
#準備（directoryの変数設定）
strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
rawdir=~/hmiwa1/rawdata/Oceania

In [71]:
cd $strdir
ls

Basedata                                         TNG.clumped
out.log                                          TNG.log
StatGenSummerSchool2023_GenomeDataAnalysis5      TNG.nosex
StatGenSummerSchool2023_GenomeDataAnalysis5.zip  TNG.valid.snp
targetdata


In [72]:
wget https://github.com/choishingwan/PRSice/releases/download/2.3.3/PRSice_linux.zip
unzip PRSice_linux.zip
ls

--2023-12-25 13:37:29--  https://github.com/choishingwan/PRSice/releases/download/2.3.3/PRSice_linux.zip
Resolving github.com (github.com)... 20.27.177.113
Connecting to github.com (github.com)|20.27.177.113|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/64860041/1e8b8800-d714-11ea-830f-cf872a4c01fe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20231225%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20231225T043730Z&X-Amz-Expires=300&X-Amz-Signature=643ac8b03383020c9669d4f3604ad1b763354924284c1f01e6e969193c09b54f&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=64860041&response-content-disposition=attachment%3B%20filename%3DPRSice_linux.zip&response-content-type=application%2Foctet-stream [following]
--2023-12-25 13:37:30--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/64860041/1e8b8800-d714-11ea-830f-cf872a4c01fe?X-Amz

In [76]:
ls $gitdir
column -s , -t $gitdir/Oceania_Tonga_231223.csv | head -n 10

05_231211_RePRSTrials.ipynb  Oceania_Tonga_231223.xlsx
Documents                    out.log
het.R                        pic
jupyter_venv5                test.ipynb
Oceania_all_170711.xlsx      Tongan.chrX.QC.FIXP.phased.log
Oceania_Tonga_231223.csv


IDNO  AGE  SEX  Height  Weight  BMI
T3    53   0    167.5   73.2    26.09044331
T104  50   0    173.4   107.9   35.88585173
T105  55   0    179.1   131.8   41.08887386
T106  55   0    172.2   109.3   36.85987582
T109  21   0    171.5   78      26.51956243
T120  47   0    175.9   92      29.73419248
T129  51   0    177.5   102     32.37452886
T134  36   0    177.1   100.4   32.0108046
T150  42   0    170.3   104.9   36.16980707


In [101]:
#mv Tonga_BMI Tonga_BMI.txt
echo "FID IID BMI" > Tonga_BMI.txt
column -s , -t $gitdir/Oceania_Tonga_231223.csv | awk '{print $1,$1,$6}' | grep -v IDNO >> Tonga_BMI.txt

In [102]:
head Tonga_BMI.txt

FID IID BMI
T3 T3 26.09044331
T104 T104 35.88585173
T105 T105 41.08887386
T106 T106 36.85987582
T109 T109 26.51956243
T120 T120 29.73419248
T129 T129 32.37452886
T134 T134 32.0108046
T150 T150 36.16980707


TNGの主成分分析結果が必要らしいので。

In [78]:
plink \
    --bfile targetdata/TNG.QC \
    --extract targetdata/TNG.QC.snplist.rs \
    --pca \
    --out targetdata/TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to targetdata/TNG.QC.log.
Options in effect:
  --bfile targetdata/TNG.QC
  --extract targetdata/TNG.QC.snplist.rs
  --out targetdata/TNG.QC
  --pca

257564 MB RAM detected; reserving 128782 MB for main workspace.
8708432 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to targetdata/TNG.QC.nosex .
--extract: 8612836 variants remaining.
Using up to 63 threads (change this with --threads).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
8612836 variants and 31 people pass filters and

In [81]:
head -n 3 targetdata/TNG.QC.eigenvec

T03 T03 0.157955 -0.1285 0.881719 -0.371433 -0.0088364 -0.0127638 -0.0261872 -0.0260119 0.0684723 -0.00758487 -0.0369127 -0.0066699 0.00950823 0.005936 -0.00417458 0.00521888 0.0264251 0.00937126 -0.000621784 0.0215547
T104 T104 0.0682922 -0.0519408 -0.0217664 -0.00428438 -0.0151739 -0.0494871 0.0352017 0.143588 -0.221894 -0.043567 -0.0384363 -0.150855 -0.35695 -0.280914 -0.05657 -0.184435 0.220318 0.209132 0.3395 0.144822
T105 T105 0.102476 -0.0587434 -0.066612 0.0446919 -0.0657773 -0.0545222 0.00434261 0.034293 -0.147696 0.0534711 0.0971919 -0.0951043 -0.0547342 0.0397273 0.156802 0.213939 0.674224 0.179705 -0.353924 0.18762


In [83]:
echo "FID IID Sex PC1 PC2 PC3 PC4 PC5 PC6" > TNG.covariate
#性別はすべて男性なので１
awk '{print $1,$2,1,$3,$4,$5,$6,$7,$8}' targetdata/TNG.QC.eigenvec >> TNG.covariate
head -n 3 TNG.covariate

FID IID Sex PC1 PC2 PC3 PC4 PC5 PC6
T03 T03 1 0.157955 -0.1285 0.881719 -0.371433 -0.0088364 -0.0127638
T104 T104 1 0.0682922 -0.0519408 -0.0217664 -0.00428438 -0.0151739 -0.0494871


In [85]:
ls Basedata/*

Basedata/BMI_Eastern_Asian.fmt.gzip    Basedata/BMI_Eastern_Asian.QC2.txt
Basedata/BMI_Eastern_Asian.newcol.txt  Basedata/BMI_Eastern_Asian.QC.txt
Basedata/BMI_Eastern_Asian.nodup.txt   Basedata/BMI_Eastern_Asian.txt


In [87]:
ls targetdata/TNG*

targetdata/TNG.bed           targetdata/TNG.QC.snplist
targetdata/TNG.bim           targetdata/TNG.QC.snplist.rs
targetdata/TNG.fam           targetdata/TNG.valid.sample
targetdata/TNG.log           targetdata/TNG.X.bed
targetdata/TNG.nosex         targetdata/TNG.X.bim
targetdata/TNG.QC.bed        targetdata/TNG.X.fam
targetdata/TNG.QC.bim        targetdata/TNG.X.log
targetdata/TNG.QC.eigenval   targetdata/TNG.X.nosex
targetdata/TNG.QC.eigenvec   targetdata/TNG.X.QC.fam
targetdata/TNG.QC.fam        targetdata/TNG.X.QC.log
targetdata/TNG.QC.het        targetdata/TNG.X.QC.nosex
targetdata/TNG.QC.log        targetdata/TNG.X.QC.prune.in
targetdata/TNG.QC.nosex      targetdata/TNG.X.QC.prune.out
targetdata/TNG.QC.prune.in   targetdata/TNG.X.QC.sexcheck
targetdata/TNG.QC.prune.out  targetdata/TNG.X.QC.snplist
targetdata/TNG.QC.rel.id


PRSice-2の実行

In [104]:
Rscript PRSice.R \
    --prsice PRSice_linux \
    --base Basedata/BMI_Eastern_Asian.QC2.txt \
    --target targetdata/TNG.QC \
    --binary-target F \
    --pheno Tonga_BMI.txt \
    --cov TNG.covariate \
    --stat beta \
    --A1 REF \
    --A2 ALT \
    --beta \
    --snp SNPNAME \
    --pvalue Pvalue \
    --out TNG

PRSice 2.3.3 (2020-08-05) 
https://github.com/choishingwan/PRSice
(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly
GNU General Public License v3
If you use PRSice in any published work, please cite:
Choi SW, O'Reilly PF.
PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.
GigaScience 8, no. 7 (July 1, 2019)
2023-12-25 14:08:03
./PRSice_linux \
    --a1 REF \
    --a2 ALT \
    --bar-levels 0.001,0.05,0.1,0.2,0.3,0.4,0.5,1 \
    --base Basedata/BMI_Eastern_Asian.QC2.txt \
    --beta  \
    --binary-target F \
    --chr CHR \
    --clump-kb 250kb \
    --clump-p 1.000000 \
    --clump-r2 0.100000 \
    --cov TNG.covariate \
    --interval 5e-05 \
    --lower 5e-08 \
    --num-auto 22 \
    --out TNG \
    --pheno Tonga_BMI.txt \
    --pvalue Pvalue \
    --seed 148255179 \
    --snp SNPNAME \
    --stat beta \
    --target targetdata/TNG.QC \
    --thread 1 \
    --upper 0.5

Initializing Genotype file: targetdata/TNG.QC (bed) 

Start processing BMI_Eastern_Asian.QC2 


In [106]:
ls TNG*

TNG_BARPLOT_2023-12-25.png  TNG_HIGH-RES_PLOT_2023-12-25.png  TNG.summary
TNG.best                    TNG.log                           TNG.valid.snp
TNG.clumped                 TNG.nosex
TNG.covariate               TNG.prsice


In [107]:
head TNG.best

FID IID In_Regression PRS
T03 T03 No 0.0615
T104 T104 Yes 0.0355
T105 T105 Yes 0.053625
T106 T106 Yes 0.04425
T109 T109 Yes 0.0545
T120 T120 Yes 0.05125
T129 T129 Yes 0.036375
T134 T134 Yes 0.046625
T150 T150 Yes 0.043375


In [108]:
head TNG.prsice

Pheno	Set	Threshold	R2	P	Coefficient	Standard.Error	Num_SNP
-	Base	0.00020005	0.00316695	0.787074	-30.7626	112.874	1
-	Base	0.00025005	0.0235325	0.458432	-100.02	133.161	2
-	Base	0.00040005	0.124066	0.0775607	-270.45	147.975	4
-	Base	0.00045005	0.0161013	0.540715	-121.155	195.788	6
-	Base	0.00055005	8.03032e-05	0.965728	9.10779	210.209	8
-	Base	0.00060005	0.00034599	0.928927	18.4686	205.327	9
-	Base	0.00080005	0.011239	0.609883	112.296	217.778	11
-	Base	0.00085005	0.0162431	0.538918	139.348	224.186	12
-	Base	0.00100005	0.016145	0.54016	144.731	233.566	14


In [109]:
head TNG.summary

Phenotype	Set	Threshold	PRS.R2	Full.R2	Null.R2	Prevalence	Coefficient	Standard.Error	P	Num_SNP
-	Base	0.00040005	0.124066	0.182896	0.0588305	-	-270.45	147.975	0.0775607	4


In [110]:
ls $gitdir

05_231211_RePRSTrials.ipynb  Oceania_Tonga_231223.xlsx
Documents                    out.log
het.R                        pic
jupyter_venv5                test.ipynb
Oceania_all_170711.xlsx      Tongan.chrX.QC.FIXP.phased.log
Oceania_Tonga_231223.csv


In [111]:
cp *.png $gitdir/pic/

In [112]:
ls $gitdir/pic/

1KG_EUR-RAGWAS_BARPLOT_2023-12-11.png        1KG_EUR-RAGWAS-PRSboxplot.png
1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        TNG_BARPLOT_2023-12-25.png
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-11.png  TNG_HIGH-RES_PLOT_2023-12-25.png
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png


<img src="pic/TNG_BARPLOT_2023-12-25.png" width=49%> <img src="pic/TNG_HIGH-RES_PLOT_2023-12-25.png" width=49%>

これでいったんそれらしきものは出た（がこれで良いのだろうか…）

混血個体を抜いてやってみた方が良いかもしれない。

### 混血個体を抜いてPRS

In [113]:
#準備（directoryの変数設定）
strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
rawdir=~/hmiwa1/rawdata/Oceania

In [114]:
ls

Basedata
lib
out.log
PRSice_linux
PRSice_linux.zip
PRSice.R
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip
targetdata
TNG_BARPLOT_2023-12-25.png
TNG.best
TNG.clumped
TNG.covariate
TNG_HIGH-RES_PLOT_2023-12-25.png
TNG.log
TNG.nosex
TNG.prsice
TNG.summary
TNG.valid.snp
Tonga_BMI.txt
TOY_BASE_GWAS.assoc
TOY_TARGET_DATA.bed
TOY_TARGET_DATA.bim
TOY_TARGET_DATA.fam
TOY_TARGET_DATA.pheno


混血個体の除外

In [115]:
cd $strdir/targetdata
echo "T03" > remove_ids.txt
echo "T129" >> remove_ids.txt
echo "T134" >> remove_ids.txt
echo "T156" >> remove_ids.txt
echo "T33" >> remove_ids.txt
echo "T47" >> remove_ids.txt
echo "T73" >> remove_ids.txt
echo "T74" >> remove_ids.txt
echo "t51" >> remove_ids.txt

In [119]:
head remove_ids.txt

T03
T129
T134
T156
T33
T47
T73
T74
t51


In [123]:
ls TNG*

TNG.bed           TNG.QC.eigenval   TNG.QC.snplist     TNG.X.QC.log
TNG.bim           TNG.QC.eigenvec   TNG.QC.snplist.rs  TNG.X.QC.nosex
TNG.fam           TNG.QC.fam        TNG.valid.sample   TNG.X.QC.prune.in
TNG.log           TNG.QC.het        TNG.X.bed          TNG.X.QC.prune.out
TNG.nonmix.log    TNG.QC.log        TNG.X.bim          TNG.X.QC.sexcheck
TNG.nonmix.nosex  TNG.QC.nosex      TNG.X.fam          TNG.X.QC.snplist
TNG.nosex         TNG.QC.prune.in   TNG.X.log
TNG.QC.bed        TNG.QC.prune.out  TNG.X.nosex
TNG.QC.bim        TNG.QC.rel.id     TNG.X.QC.fam


In [126]:
cat TNG.fam | grep -v -f remove_ids.txt > keep_ids.txt

In [127]:
plink \
    --vcf Tongan.QC.FIXP.phased.annotated.vcf.gz \
    --make-bed \
    --keep keep_ids.txt \
    --out TNG.nonmix

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.nonmix.log.
Options in effect:
  --keep keep_ids.txt
  --make-bed
  --out TNG.nonmix
  --vcf Tongan.QC.FIXP.phased.annotated.vcf.gz

257564 MB RAM detected; reserving 128782 MB for main workspace.
--vcf: TNG.nonmix-temporary.bed + TNG.nonmix-temporary.bim +
TNG.nonmix-temporary.fam written.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nonmix.nosex .
--keep: 22 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 22 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genot

In [129]:
ls TNG.nonmix*

TNG.nonmix.bed  TNG.nonmix.fam  TNG.nonmix.nosex
TNG.nonmix.bim  TNG.nonmix.log


QCのやり直し

In [130]:
plink \
    --bfile TNG.nonmix \
    --maf 0.01 \
    --hwe 1e-6 \
    --geno 0.01 \
    --mind 0.01 \
    --write-snplist \
    --make-just-fam \
    --out TNG.nonmix.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.nonmix.QC.log.
Options in effect:
  --bfile TNG.nonmix
  --geno 0.01
  --hwe 1e-6
  --maf 0.01
  --make-just-fam
  --mind 0.01
  --out TNG.nonmix.QC
  --write-snplist

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
22 people (0 males, 0 females, 22 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nonmix.QC.nosex .
0 people removed due to missing genotype data (--mind).
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 22 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
0 var

In [131]:
#プルーニング
plink \
    --bfile TNG.nonmix \
    --keep TNG.nonmix.QC.fam \
    --extract TNG.nonmix.QC.snplist \
    --indep-pairwise 200 50 0.25 \
    --out TNG.nonmix.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.nonmix.QC.log.
Options in effect:
  --bfile TNG.nonmix
  --extract TNG.nonmix.QC.snplist
  --indep-pairwise 200 50 0.25
  --keep TNG.nonmix.QC.fam
  --out TNG.nonmix.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
22 people (0 males, 0 females, 22 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nonmix.QC.nosex .
--extract: 7769431 variants remaining.
--keep: 22 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 22 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is ex

In [132]:
#ヘテロ接合度チェック
plink \
    --bfile TNG.nonmix \
    --extract TNG.nonmix.QC.prune.in \
    --keep TNG.nonmix.QC.fam \
    --het \
    --out TNG.nonmix.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.nonmix.QC.log.
Options in effect:
  --bfile TNG.nonmix
  --extract TNG.nonmix.QC.prune.in
  --het
  --keep TNG.nonmix.QC.fam
  --out TNG.nonmix.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
22 people (0 males, 0 females, 22 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nonmix.QC.nosex .
--extract: 451789 variants remaining.
--keep: 22 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 22 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
451789 variant

In [135]:
cd $strdir
cat << EOL > $gitdir/het2.R
setwd("$strdir/targetdata")
dat <- read.table("TNG.nonmix.QC.het", header=T) # Read in the EUR.het file, specify it has header
m <- mean(dat\$F) # Calculate the mean
s <- sd(dat\$F) # Calculate the SD
valid <- subset(dat, F <= m+3*s & F >= m-3*s) # Get any samples with F coefficient within 3 SD of the population mean
write.table(valid[,c(1,2)], "TNG.nonmix.valid.sample", quote=F, row.names=F) # print FID and IID for valid samples
EOL
cat $gitdir/het2.R

setwd("/home/hmiwa/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials/targetdata")
dat <- read.table("TNG.nonmix.QC.het", header=T) # Read in the EUR.het file, specify it has header
m <- mean(dat$F) # Calculate the mean
s <- sd(dat$F) # Calculate the SD
valid <- subset(dat, F <= m+3*s & F >= m-3*s) # Get any samples with F coefficient within 3 SD of the population mean
write.table(valid[,c(1,2)], "TNG.nonmix.valid.sample", quote=F, row.names=F) # print FID and IID for valid samples


In [136]:
Rscript --no-save $gitdir/het2.R

In [140]:
ls TNG.nonmix*

TNG.nonmix.bed    TNG.nonmix.QC.fam       TNG.nonmix.QC.prune.out
TNG.nonmix.bim    TNG.nonmix.QC.het       TNG.nonmix.QC.snplist
TNG.nonmix.fam    TNG.nonmix.QC.log       TNG.nonmix.valid.sample
TNG.nonmix.log    TNG.nonmix.QC.nosex
TNG.nonmix.nosex  TNG.nonmix.QC.prune.in


In [143]:
wc TNG.nonmix.fam
wc TNG.nonmix.valid.sample
head TNG.nonmix.valid.sample

 22 132 396 TNG.nonmix.fam


 23  46 206 TNG.nonmix.valid.sample
FID IID
T104 T104
T105 T105
T106 T106
T109 T109
T120 T120
T150 T150
T153 T153
T15 T15
T26 T26


In [146]:
#relatednessチェック
plink \
    --bfile TNG.nonmix \
    --keep TNG.nonmix.valid.sample \
    --rel-cutoff 0.125 \
    --out TNG.nonmix.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.nonmix.QC.log.
Options in effect:
  --bfile TNG.nonmix
  --keep TNG.nonmix.valid.sample
  --out TNG.nonmix.QC
  --rel-cutoff 0.125

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
22 people (0 males, 0 females, 22 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nonmix.QC.nosex .
--keep: 22 people remaining.
Using up to 63 threads (change this with --threads).
Before main variant filters, 22 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
8711185 variants and 22 people pass filters and QC (before --rel-cutoff

In [147]:
#サンプルはすべて使えそう
plink \
    --bfile TNG.nonmix \
    --make-bed \
    --keep TNG.nonmix.QC.fam \
    --extract TNG.nonmix.QC.snplist \
    --out TNG.nonmix.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.nonmix.QC.log.
Options in effect:
  --bfile TNG.nonmix
  --extract TNG.nonmix.QC.snplist
  --keep TNG.nonmix.QC.fam
  --make-bed
  --out TNG.nonmix.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
22 people (0 males, 0 females, 22 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nonmix.QC.nosex .
--extract: 7769431 variants remaining.
--keep: 22 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 22 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
7769431 v

In [151]:
cd $strdir
ls TNG*

TNG_BARPLOT_2023-12-25.png  TNG_HIGH-RES_PLOT_2023-12-25.png  TNG.summary
TNG.best                    TNG.log                           TNG.valid.snp
TNG.clumped                 TNG.nosex
TNG.covariate               TNG.prsice


In [152]:
cd $strdir
ls targetdata/TNG*

targetdata/TNG.bed                  targetdata/TNG.QC.eigenvec
targetdata/TNG.bim                  targetdata/TNG.QC.fam
targetdata/TNG.fam                  targetdata/TNG.QC.het
targetdata/TNG.log                  targetdata/TNG.QC.log
targetdata/TNG.nonmix.bed           targetdata/TNG.QC.nosex
targetdata/TNG.nonmix.bim           targetdata/TNG.QC.prune.in
targetdata/TNG.nonmix.fam           targetdata/TNG.QC.prune.out
targetdata/TNG.nonmix.log           targetdata/TNG.QC.rel.id
targetdata/TNG.nonmix.nosex         targetdata/TNG.QC.snplist
targetdata/TNG.nonmix.QC.bed        targetdata/TNG.QC.snplist.rs
targetdata/TNG.nonmix.QC.bim        targetdata/TNG.valid.sample
targetdata/TNG.nonmix.QC.fam        targetdata/TNG.X.bed
targetdata/TNG.nonmix.QC.het        targetdata/TNG.X.bim
targetdata/TNG.nonmix.QC.log        targetdata/TNG.X.fam
targetdata/TNG.nonmix.QC.nosex      targetdata/TNG.X.log
targetdata/TNG.nonmix.QC.prune.in   targetdata/TNG.X.nosex
targetdata/TNG.nonmix.QC.prune.out  t

In [154]:
cat targetdata/TNG.nonmix.QC.snplist | wc
cat targetdata/TNG.nonmix.QC.snplist | grep rs | wc
cat targetdata/TNG.nonmix.QC.snplist | grep rs > targetdata/TNG.nonmix.QC.snplist.rs

7743217 7743217 83315660
7673835 7673835 83176896


Clumping

In [155]:
cd $strdir
plink \
    --bfile targetdata/TNG.nonmix.QC \
    --extract targetdata/TNG.nonmix.QC.snplist.rs \
    --clump-p1 1 \
    --clump-r2 0.1 \
    --clump-kb 250 \
    --clump Basedata/BMI_Eastern_Asian.QC2.txt \
    --clump-snp-field SNPNAME \
    --clump-field Pvalue \
    --out TNG.nonmix

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.nonmix.log.
Options in effect:
  --bfile targetdata/TNG.nonmix.QC
  --clump Basedata/BMI_Eastern_Asian.QC2.txt
  --clump-field Pvalue
  --clump-kb 250
  --clump-p1 1
  --clump-r2 0.1
  --clump-snp-field SNPNAME
  --extract targetdata/TNG.nonmix.QC.snplist.rs
  --out TNG.nonmix

257564 MB RAM detected; reserving 128782 MB for main workspace.
7769431 variants loaded from .bim file.
22 people (0 males, 0 females, 22 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nonmix.nosex .
--extract: 7673835 variants remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 22 founders and 0 nonfounders present.
Calculating allele frequencies... 1011121314151617181920212223242526272829303132333435363738394041424344454647484950515253545556575859606162636465666768697071727374757677

In [156]:
awk 'NR!=1{print $3}' TNG.nonmix.clumped >  TNG.valid.snp

PRSice-2を動かします

In [157]:
#準備（directoryの変数設定）
strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
rawdir=~/hmiwa1/rawdata/Oceania

主成分分析結果用意

In [158]:
plink \
    --bfile targetdata/TNG.nonmix.QC \
    --extract targetdata/TNG.nonmix.QC.snplist.rs \
    --pca \
    --out targetdata/TNG.nonmix.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to targetdata/TNG.nonmix.QC.log.
Options in effect:
  --bfile targetdata/TNG.nonmix.QC
  --extract targetdata/TNG.nonmix.QC.snplist.rs
  --out targetdata/TNG.nonmix.QC
  --pca

257564 MB RAM detected; reserving 128782 MB for main workspace.
7769431 variants loaded from .bim file.
22 people (0 males, 0 females, 22 ambiguous) loaded from .fam.
Ambiguous sex IDs written to targetdata/TNG.nonmix.QC.nosex .
--extract: 7673835 variants remaining.
Using up to 63 threads (change this with --threads).
Before main variant filters, 22 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
7673835 vari

In [159]:
echo "FID IID Sex PC1 PC2 PC3 PC4 PC5 PC6" > TNG.nonmix.covariate
#性別はすべて男性なので１
awk '{print $1,$2,1,$3,$4,$5,$6,$7,$8}' targetdata/TNG.nonmix.QC.eigenvec >> TNG.nonmix.covariate
head -n 3 TNG.nonmix.covariate

FID IID Sex PC1 PC2 PC3 PC4 PC5 PC6
T104 T104 1 0.117274 0.129155 0.167074 0.24903 -0.386582 -0.0551921
T105 T105 1 0.173793 -0.0668831 0.088268 0.0464156 0.0955457 0.0472691


In [160]:
Rscript PRSice.R \
    --prsice PRSice_linux \
    --base Basedata/BMI_Eastern_Asian.QC2.txt \
    --target targetdata/TNG.nonmix.QC \
    --binary-target F \
    --pheno Tonga_BMI.txt \
    --cov TNG.nonmix.covariate \
    --stat beta \
    --A1 REF \
    --A2 ALT \
    --beta \
    --snp SNPNAME \
    --pvalue Pvalue \
    --out TNG.nonmix

PRSice 2.3.3 (2020-08-05) 
https://github.com/choishingwan/PRSice
(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly
GNU General Public License v3
If you use PRSice in any published work, please cite:
Choi SW, O'Reilly PF.
PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.
GigaScience 8, no. 7 (July 1, 2019)
2023-12-25 18:42:28
./PRSice_linux \
    --a1 REF \
    --a2 ALT \
    --bar-levels 0.001,0.05,0.1,0.2,0.3,0.4,0.5,1 \
    --base Basedata/BMI_Eastern_Asian.QC2.txt \
    --beta  \
    --binary-target F \
    --chr CHR \
    --clump-kb 250kb \
    --clump-p 1.000000 \
    --clump-r2 0.100000 \
    --cov TNG.nonmix.covariate \
    --interval 5e-05 \
    --lower 5e-08 \
    --num-auto 22 \
    --out TNG.nonmix \
    --pheno Tonga_BMI.txt \
    --pvalue Pvalue \
    --seed 2673475949 \
    --snp SNPNAME \
    --stat beta \
    --target targetdata/TNG.nonmix.QC \
    --thread 1 \
    --upper 0.5

Initializing Genotype file: targetdata/TNG.nonmix.QC (bed) 

Start proce

In [161]:
ls TNG.nonmix*

TNG.nonmix_BARPLOT_2023-12-25.png        TNG.nonmix.log
TNG.nonmix.best                          TNG.nonmix.nosex
TNG.nonmix.clumped                       TNG.nonmix.prsice
TNG.nonmix.covariate                     TNG.nonmix.summary
TNG.nonmix_HIGH-RES_PLOT_2023-12-25.png


In [162]:
head TNG.nonmix.best

FID IID In_Regression PRS
T104 T104 Yes 0.0297188865
T105 T105 Yes 0.0311173581
T106 T106 Yes 0.0303362445
T109 T109 Yes 0.0300234716
T120 T120 Yes 0.0302407205
T150 T150 Yes 0.030415393
T153 T153 Yes 0.0307767467
T15 T15 Yes 0.0303034934
T26 T26 Yes 0.0309519651


In [163]:
head TNG.nonmix.prsice

Pheno	Set	Threshold	R2	P	Coefficient	Standard.Error	Num_SNP
-	Base	0.00020005	0.0386394	0.398329	127.327	147.821	1
-	Base	0.00025005	0.00933942	0.682004	-100.487	242.011	2
-	Base	0.00040005	0.0573311	0.299341	-291.144	273.906	4
-	Base	0.00045005	0.00245597	0.834085	58.2495	274.807	6
-	Base	0.00055005	0.0220196	0.526898	163.133	253.723	8
-	Base	0.00060005	0.0467118	0.35126	265.698	278.988	9
-	Base	0.00080005	0.0621577	0.278752	313.442	282.24	11
-	Base	0.00085005	0.0584057	0.29461	327.978	305.477	12
-	Base	0.00100005	0.0430128	0.371891	322.206	353.474	14


In [164]:
head TNG.nonmix.summary

Phenotype	Set	Threshold	PRS.R2	Full.R2	Null.R2	Prevalence	Coefficient	Standard.Error	P	Num_SNP
-	Base	0.0980001	0.154568	0.386832	0.232263	-	10496.1	5587.21	0.0736153	916


In [166]:
cp *.png $gitdir/pic/
ls $gitdir/pic/

1KG_EUR-RAGWAS_BARPLOT_2023-12-11.png
1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-11.png
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png
1KG_EUR-RAGWAS-PRSboxplot.png
TNG_BARPLOT_2023-12-25.png
TNG_HIGH-RES_PLOT_2023-12-25.png
TNG.nonmix_BARPLOT_2023-12-25.png
TNG.nonmix_HIGH-RES_PLOT_2023-12-25.png


<img src="pic/TNG.nonmix_BARPLOT_2023-12-25.png" width=49%> <img src="pic/TNG.nonmix_HIGH-RES_PLOT_2023-12-25.png" width=49%>

とりあえずPRSの算出まで動かすことに成功した。

これ、他の集団との分布（全体的に大きいとか小さいとか）がわかったほうがいいんだろうな。→使えそうなサンプルがないか探してみるか、ランダムに設定したものと比較した場合でみるとかになるのかな。東アジアのデータが使えると良いのだけれども？

過去のPRSを推定する方法を実際にやっていく必要がある

ポリジェニックセレクションについても前読んだジャーナルから進めていかなければ。